File: src/main/java/ar/edu/unq/spring/controller/dto/PersonajeDTO.java
package ar.edu.unq.spring.controller.dto;

import ar.edu.unq.spring.modelo.Personaje;
import lombok.Getter;
import lombok.Setter;

import java.util.HashSet;
import java.util.Set;

@Getter @Setter
public class PersonajeDTO {

    public final Long id;
    public final String nombre;
    public final int vida;
    public final int pesoMaximo;
    public final Set<ItemDTO> inventario;

    public PersonajeDTO(Long id, String nombre, int vida, int pesoMaximo, Set<ItemDTO> inventario) {
        this.id = id;
        this.nombre = nombre;
        this.vida = vida;
        this.pesoMaximo = pesoMaximo;
        this.inventario = inventario;
    }

    public static PersonajeDTO desdeModelo(Personaje personaje) {
        Set<ItemDTO> inventarioDTO = personaje.getInventario() != null ?
                new HashSet<>(personaje.getInventario().stream()
                        .map(ItemDTO::desdeModelo)
                        .toList()) : null;

        return new PersonajeDTO(
                personaje.getId(),
                personaje.getNombre(),
                personaje.getVida(),
                personaje.getPesoMaximo(),
                inventarioDTO
        );
    }

    public Personaje aModelo() {
        Personaje personaje = new Personaje();

        personaje.setId(this.id);
        personaje.setNombre(this.nombre);
        personaje.setVida(this.vida);
        personaje.setPesoMaximo(this.pesoMaximo);
        personaje.setInventario(inventario != null ?
                new HashSet<>(inventario.stream()
                        .map(itemDTO -> itemDTO.aModelo(personaje))
                        .toList()) : new HashSet<>());

        return personaje;
    }
}

File: src/main/kotlin/ar/edu/unq/spring/controller/dto/PersonajeDTO.kt
package ar.edu.unq.spring.controller.dto

import ar.edu.unq.spring.modelo.Personaje
import ar.edu.unq.spring.modelo.clasesDePersonajes.Guerrero
import ar.edu.unq.spring.modelo.clasesDePersonajes.Mago
import ar.edu.unq.spring.modelo.clasesDePersonajes.Picaro

class PersonajeDTO(
    val id: Long?,
    val nombre: String?,
    val vida: Int,
    val pesoMaximo: Int,
    val tipo: String,
    val estadisticaDeClase: Int,
    val inventario: MutableSet<ItemDTO>?
) {
    companion object {
        fun desdeModelo(personaje: Personaje) =
            PersonajeDTO(
                id = personaje.id,
                nombre = personaje.nombre,
                vida = personaje.vida,
                pesoMaximo = personaje.pesoMaximo,
                tipo = when (personaje) {
                    is Guerrero -> "Guerrero"
                    is Mago -> "Mago"
                    is Picaro -> "Picaro"
                    else -> throw IllegalArgumentException("No existe esa subclase")
                },
                estadisticaDeClase = when (personaje) {
                    is Guerrero -> (personaje as Guerrero).fuerza
                    is Mago -> (personaje as Mago).mana
                    is Picaro -> (personaje as Picaro).sigilo
                    else -> throw IllegalArgumentException("No existe esa subclase")
                },
                inventario = personaje.inventario
                    ?.map { item -> ItemDTO.desdeModelo(item) }
                    ?.toCollection(HashSet())
            )
    }

    fun aModelo(): Personaje {
        val personaje: Personaje = when (tipo) {
            "Guerrero" -> Guerrero().apply { fuerza = estadisticaDeClase }
            "Mago" -> Mago().apply { mana = estadisticaDeClase }
            "Picaro" -> Picaro().apply { sigilo = estadisticaDeClase }
            else -> throw IllegalArgumentException("No existe esa subclase")
        }
        personaje.id = this.id
        personaje.nombre = this.nombre
        personaje.vida = this.vida
        personaje.pesoMaximo = this.pesoMaximo
        personaje.inventario = this.inventario
            ?.map { itemDTO -> itemDTO.aModelo(personaje) }
            ?.toCollection(HashSet())
            ?: HashSet()
        return personaje
    }
}
